### ISSUE TITLE
Steps to build ios bridge need some clarification

### ISSUE THREAD

**Issue by vividcode** [scapix-com/scapix#17]
Hey! First off, great project! 
I find this tool quite simplistic compared to others out there to bridge c++ with mobile platform.

However I have trouble understanding the steps involved (maybe I am not very well versed in cmake).

While example projects are great, it would be desirable to have step by step instruction about how one goes on to build something from scratch? This is because example projects do not tell what is generated vs what is bare minimum.

Here is my CMakeList.txt:

```
cmake_minimum_required(VERSION 3.14)

project(MyProject)

include(FetchContent)
FetchContent_Declare(
  cmodule
  URL "https://github.com/scapix-com/cmodule/archive/v1.0.22.tar.gz"
  URL_HASH SHA256=f331d227ae1be52677833baf4fd4728a7c6ee66ec570db6e60d22db0ae43854f
)
FetchContent_MakeAvailable(cmodule)

find_package(Scapix REQUIRED)

scapix_add_target(
    mylib
    "com.example.mylib"
    "cppSource/myclass.hpp"
    "cppSource/myclass.cpp"
)
```
Here, MyProject is my desired project name and mylib is my intended library name (in iOS target parlance).

Without _project_ line, it complains about it being required. When I supply it to be my desired MyProject, it complains it is not there. But I see sometimes it generates the project automatically.

The command that I run is `bash xcode_ios_objc.sh` while I encounter project missing error. Previous error (probably the cause of it) is:

```
CMake Error at /../CMakeLists.txt:492 (add_custom_command):
  Attempt to add a custom rule to output

    /Users/Admin/Desktop/Mapbox/MapProject/generated/bridge/java/map.cpp.rule

  which already has a custom rule.
Call Stack (most recent call first):
  /Users/Admin/.cmodule/scapix/1.0.20/CMakeLists.txt:104 (scapix_bridge)
  /Users/Admin/.cmodule/scapix/1.0.20/CMakeLists.txt:518 (scapix_bridge_headers)
  CMakeLists.txt:19 (scapix_add_target)
```

Again, while I understand my knowledge of CMake is not great, I think some steps (in readme or in examples) how one goes about it generating project from scratch would be greatly helpful.

Thanks again for the most useful utility!

**Comment by Boris-Rasin** [comment]
Thanks!

Here is a new minimal example specifically for XCode:
https://github.com/scapix-com/scapix_example_objc

The readme also contains list of steps used to create project from scratch.
Please tell me how this could be improved further.

**Comment by vividcode** [comment]
Thanks for quick follow up on my request!

Just checked it, it is all building fine as I run `cmake_generate.sh`. 

Some things are I noticed: (I do not want to clutter it with new issues so continuing with this one)

- Past running .sh file, every target is generated for Mac. Should there be a target generated for iOS?
- The end goal is to generate a library that can be included in client application - not sure what target should be built?
- As I tried building cpplib scheme, I got many errors in my C++17 code. I assume it's some setting the project maybe missing. I am searching for that.

Thanks again for the quick follow up on the request.

**Comment by vividcode** [comment]
Also - Looking at the previous repo, was there any hint of problem with my CMakeList that you could spot?
Maybe it's more general approach but I am struggling in `scapix_add_target` command:

```
cmake_minimum_required(VERSION 3.14)

project(MyProject CXX)

include(FetchContent)
FetchContent_Declare(
  cmodule
  URL "https://github.com/scapix-com/cmodule/archive/v1.0.22.tar.gz"
  URL_HASH SHA256=f331d227ae1be52677833baf4fd4728a7c6ee66ec570db6e60d22db0ae43854f
)
FetchContent_MakeAvailable(cmodule)

set(SCAPIX_BRIDGE "java" CACHE STRING "cpp, java, objc, python, js, cs")
set(SCAPIX_PLATFORM "windows" CACHE STRING "one of the folders inside 'source/chat/platform': ios, android, macos, windows, linux, etc.")
# # set(SCAPIX_JAVA_API "android-28" CACHE STRING "one of the folders inside 'scapix/java_api': jdk-11.0.2, android-28, etc.")

find_package(Scapix REQUIRED)

scapix_add_target(
    MyProject
    "com.iphonegamezone.MyProject"
    "cppSource/my.hpp"
    "cppSource/my.cpp"
)

```

**Comment by Boris-Rasin** [comment]
`Steps to integrate generated cpplib are probably redundant in readme (or labeled as what the bash script does), as I saw they were already done past running .sh file.`

I wasn't clear about this, these steps were done manually to integrate generated C++ project to existing ObjC project. You will have to repeat these steps to integrate to your own existing ObjC project. I have updated the readme to clarify.

`Past running .sh file, every target is generated for Mac. Should there be a target generated for iOS?`

To generate target for iOS, you need to add CMake parameter `-DCMAKE_SYSTEM_NAME=iOS`, I've added [cmake_generate_ios.sh](https://github.com/scapix-com/scapix_example_objc/blob/master/cpplib/cmake_generate_ios.sh) script to demonstrate this. Such project should be integrated with ObjC/Swift project targeting iOS and will not work with `objcapp` project from example as it targets macOS.

`The end goal is to generate a library that can be included in client application - not sure what target should be built?`

Example shows how to include XCode library project as dependency of client application. When client application is built, XCode will also build it's dependencies. And, in this case, also invoke Scapix to generate bindings as part of the build. So when the library C++ API changes, new bindings are automatically generated during build.

`As I tried building cpplib scheme, I got many errors in my C++17 code. I assume it's some setting the project maybe missing. I am searching for that.`

Yes, the example `objcapp` project specifies C++14 standard, this should be changed to C++17.

`Also - Looking at the current repo (not the example you just created), was there any hint of problem with my CMakeList that you could spot?`

Only header files should be passed to `scapix_add_target`, not cpp files.

**Comment by vividcode** [comment]
Hey!

Thanks again for clarifications. I realized that you were correct, some things had to be done past it. I regenerated objcapp project for iOS and ran your new super useful cmake_generate_ios.sh.

I followed your steps to build objcapp, and it built successfully. 👍 👍 👍

However, I am  still unable to see my desired objc bridge files. 
The only .mm file I see is BridgeObject.mm which I guess is of no use.
Yet, I see the c++ library (libcpplib.a) built successfully.

See the folder structure of scapix_example.xcworkspace after I built objcapp target.

Where exactly should I look for the bridging files, and what steps I should still perform to get it if I missed something?

In other words, how do I use code from `my.cpp` and `my.hpp`?

![Screen Shot 2020-09-18 at 00 59 19](https://user-images.githubusercontent.com/3069654/93532733-3f1d4380-f94a-11ea-9a73-d08c873dc6d2.png)

**Comment by Boris-Rasin** [comment]
In the example, files are generated during build in the folder `cpplib/generated/bridge/objc/cpplib/bridge`.

There is also an example usage in `AppDelegate.m` file:
`Contact* c = [Contact new];`

**Comment by Boris-Rasin** [comment]
I have simplified XCode integration, now you only need to add generated scapix.xcconfig file to your project and it will automatically configure all settings (include paths, linked libraries, etc.).

I have also updated the example: [scapix_example_objc](https://github.com/scapix-com/scapix_example_objc).
